### PR TITLE
Create explicit warning if not building with CBT

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request
Adds a new `CBT` define which is automatically created when building. If this define is absent, the build will fail.

This is what @Cyberboss tried to do with USE_BUILD_BAT_INSTEAD_OF_DREAM_MAKER.dm, but couldn't. See https://github.com/tgstation/tgstation/pull/56470.

The reasoning behind this is CBT is already a requirement to build a fresh project, otherwise the tgui bundle files won't exist. This gives a readable error to go along with that. However, you can currently build once then just use Dream Maker. This is a footgun--not only are we already adding new things to CBT like tgfont which will fail later on, but also it will create weird scenarios when we add tasks to CBT that don't immediately fail if not ran, or otherwise create out of sync builds.

Also replaces `rmSync` with `unlinkSync`, which works on older Node versions.

### CODE REVIEW

**Comment by tralezab** [comment]
This kills the dm coder (and that's a good thing) here's why: (click here for more information)

**Comment by Mothblocks** [comment]
@Cyberboss has informed me this would break TGS. Working on it.

**Review by Cyberboss — APPROVED** [review]

**Comment by Iamgoofball** [comment]
:-1:

File: code/_compile_options.dm [hunk 1]
```diff
@@ -72,6 +72,17 @@
...
+// TGS performs its own build of dm.exe, but includes a prepended TGS define.
+#define CBT
+#endif
+
 // A reasonable number of maximum overlays an object needs
 // If you think you need more, rethink it
 #define MAX_ATOM_OVERLAYS 100
+
+#if !defined(CBT) && !defined(SPACEMAN_DMM)
+#error "Building with Dream Maker is no longer supported."
```
Comment (Iamgoofball) [inline code/_compile_options.dm:85]: unacceptable, this is simply awfulg

**Comment by Cyberboss** [comment]
Why the code is fine. I'd like to say I am rather against this, but at the same time don't care enough to block it.

**Comment by Mothblocks** [comment]
The goal is twofold:

1. To give a proper error message for when things like tgui aren't bundled, which was the original reason behind *your* PR. This way, it's immediately clear what to do.

2. Future compatibility. Take for instance the [automatic define list refactor proposition](https://hackmd.io/@tgstation/BksIEE_xO). This is something that boosts productivity a fair bit, and is something we are now able to easily keep in sync with CBT. However, it's not something that would *fail* to compile if it weren't up to date for whatever reason. For this specific example, some stuff will fail to compile (like `TRAIT_*`), but other stuff won't, and future tools might not have any obvious errors that'll show up when compiling like the tgui.bundle.js ones that exist now.

**Comment by LemonInTheDark** [comment]
I really don't like the precedent this is setting. I understand why you want this, but I don't think it's a good idea to rely this much on an external tool just to compile the game.

**Comment by Mothblocks** [comment]
We already require it to compile the game--it's just a side effect that you only need to do it once, which is an assumption that isn't actually guaranteed to hold as development pushes further.

File: code/_compile_options.dm [hunk 2]
```diff
@@ -72,6 +72,17 @@
...
+#error "In order to build, run BUILD.bat in the root directory."
+#error "Consider using Visual Studio Code instead, where you can Ctrl+Shift+B to build."
```
Comment (MrStonedOne) [inline code/_compile_options.dm:87]: Make this a warning, stop trying to stop people from making mistakes on their own computers.

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]

File: code/_compile_options.dm (hunk 2 above)
Comment (MrStonedOne) [reply code/_compile_options.dm:87]:
```suggestion
#warn "Building with Dream Maker is no longer supported."
#warn "In order to build, run BUILD.bat in the root directory."
#warn "Consider using Visual Studio Code instead, where you can Ctrl+Shift+B to build."
```

**Comment by LemonInTheDark** [comment]
> 
> 
> We already require it to compile the game--it's just a side effect that you only need to do it once, which is an assumption that isn't actually guaranteed to hold as development pushes further.

I'm not sure it's

File: code/_compile_options.dm (hunk 2 above)
Comment (Mothblocks) [reply code/_compile_options.dm:87]: Works for me, as long as it explicitly shows the footgun and they know what to do if something breaks before asking us.

**Comment by optimumtact** [comment]
This makes me nervous, in terms of adding a dependency into the main compile loop like this, but I am no longer a major contributor to the technical discussion around the tg engine, so I don't think my fears should outweigh the work of more active contributors.

ultimately then I say merge, because there appears to be genuine technical buzz from regular contributors about what this can achieve, but please be aware of the risks of adding a second busfactor to the codebase beyond lummox just getting hit by a bus and always make sure you have a backup plan to bail out of requiring CBT.

**Comment by LemonInTheDark** [comment]
Should make mention of that somewhere, so it doesn't get lost in the churn of the codebase

**Comment by stylemistake** [comment]
My feelings are also mixed, hence why I didn't do it yet.

Build tool is still very immature and full of shitcode instead of robust extensible pipelines. There must be a backup solution while this is still in a refinement process.

While I don't think a simple define that tells an error message is inherently bad (you can just remove that define in case someone is hit by a bus), relying on a build tool to auto-generate a list of includes in dme file is very bad.

DM file ticking for inclusion/exclusion in dme file is an important mechanic of developing in dm lang, a lot of tools like SpacemanDMM rely on this, refactoring this will just result in a refactoring cascade throughout all the tooling with no tangible benefits, other than "not needing to tick files" and potentially avoiding "tgstation.dme merge conficts", which are already rare because that file is sorted in filesystem order and new files appear in the middle of it, and git has no problem merging that.

TL;DR: Hold.

**Comment by Mothblocks** [comment]
> While I don't think a define with a warning message is inherently bad, relying on a build tool to auto-generate a list of includes in dme file is bad.

To be clear, the automatic define is *only* in tgstation-server, where it will never be committed anyway. This is because anything else would require modifications to TGS 3, which I have been told is too much of a challenge. It is impossible for it to merge conflict in practice.

**Review by stylemistake — APPROVED** [review]

**Comment by IndieanaJones** [comment]
Could you make sure your tool consistently works first before doing this?

**Comment by lordpidey** [comment]
Yeah, I had a HELL of a time getting the "proper" build stuff to work.

Followed the instructions.  Didn't work.

Turns out it got caught by my antivirus.  Took me a while to figure out.

**Comment by stylemistake** [comment]
@lordpidey DM me in Discord, `stylemistake#1991`, I want details.

And on that note, I'm merging this.
